### PR TITLE
normalize nested slices in Containers to be nil if slice is empty

### PR DESCRIPTION
**Pull request by hanxiaoshuai**
**What this PR does / why we need it**:
fix todo: Also normalize nested maps and slices.
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/core/equivalence_cache.go [hunk 1]
```diff
@@ -290,3 +294,33 @@ func getEquivalenceHash(pod *v1.Pod) *equivalencePod {
 	// TODO(misterikkit): Also normalize nested maps and slices.
 	return ep
 }
+
+func normalizeContainers(containers []v1.Container) {
```
Comment (wojtek-t) [inline pkg/scheduler/core/equivalence_cache.go]: I don't fully understand what is the purpose of this change. Can you please explain?

Comment (hanxiaoshuai) [reply pkg/scheduler/core/equivalence_cache.go]: Because `DeepHashObject `considers nil and empty slices to be different. So it should set to be nil if slice is empty. In func `getEquivalenceHash` it only normalize non-nestedmaps and slices. And there is TODO by @misterikkit `Also normalize nested maps and slices.` So I also normalize nested containers slices.

**Review by liggitt — COMMENTED** [review]
> Because DeepHashObject considers nil and empty slices to be different

semantic deepequals does not. I'd rather see a matching implementation of DeepHashObject if we need it… chasing mutating normalizations like this is very likely to drift and repeatedly be outdated.

**Comment by misterikkit** [comment]
@hanxiaoshuai Thanks for pitching in to help here. @wojtek-t  & @liggitt - this TODO was an artifact from my work on equivalence cache. We want to find the right balance between being formally correct and not taking too much time to compute a hash value.

When I was first changing the hash function, I naively added a test to confirm that nil and empty slice behaved the same. Rather than deleting the test, I "fixed" the code to normalize some empty/nil slices.

What I would really like to see is a faster implementation of DeepHashObject that also treats nil and empty slice the same. (Although this could easily break existing users.) I think that using spew to hash objects might be costing us.
